src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,392 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.BlockStore
+{
+    /// <summary>
+    /// Saves blocks to the database in batches, removes reorged blocks from the database.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// The batch is saved when total serialized size of all blocks in a batch reaches <see cref="BatchThresholdSizeBytes"/>,
+    /// or when more than <see cref="BatchMaxSaveIntervalSeconds"/> passed since last batch was saved, or when node is shutting down.
+    /// </para>
+    /// <para>
+    /// When we save new blocks to the database, in case <see cref="IBlockRepository"/> contains blocks that
+    /// are no longer a part of our best chain, they are removed from the database.
+    /// </para>
+    /// <para>
+    /// When block store is being initialized we delete blocks that are not on the best chain.
+    /// </para>
+    /// </remarks>
+    public class BlockStore : IDisposable
```

Comment:
No interface but I ask myself if we need one this is a high level component so wont need to be mocked away and no one will call it.

Reply:
I actually tend to avoid having interface even on low level components unless I know that another component might want to implement such an interface. 

Components without an interface can be mocked using Moq (like what I did in tests for it)

Reply:
You avoid using an interface unless another component will use it, the benefits to using an interface allow mocking it out for the top caller class.

>Or you just use the real thing in a test

if you use the real thing you need to cover it in your tests, normally each components will have there own tests 

>Components without an interface can be mocked using Moq

Can you show me an example? I was sure concrete types can not be mocked (or there was some limitation there)

Reply:
>Can you show me an example? I was sure concrete types can not be mocked (or there was some limitation there)

Yeah- there is a limitation- they should be virtual. 

Example:

```
public class TempTest
{
    [Fact]
    public void TestSomething()
    {
        var moq = new Mock<TestableClass>();
        moq.Setup(c => c.DoSomething()).Returns(delegate { return "qwe"; });

        TestableClass testable = moq.Object;
        string result = testable.DoSomething();
        Assert.Equal("qwe", result);
    }
}
public class TestableClass
{
    public virtual string DoSomething()
    {
        return "unchanged text";
    }
}
```

Reply:
And you need to provide the constructor params if there are any

Reply:
ah yes the methods need to be virtual, this allows to mock abstracts classes.